source/opt/loop_unroller.cpp

```diff
@@ -502,17 +509,17 @@ void LoopUnrollerUtilsImpl::MarkLoopControlAsDontUnroll(ir::Loop* loop) const {
   assert(loop_merge_inst &&
          "Loop merge instruction could not be found after entering unroller "
          "(should have exited before this)");
-
-  uint32_t loop_control_dont_unroll = 2;
-  uint32_t loop_control_index = 2;
-  loop_merge_inst->SetInOperand(loop_control_index, {loop_control_dont_unroll});
+  loop_merge_inst->SetInOperand(kLoopControlIndex,
+                                {kLoopControlDontUnrollIndex});
 }
 
 // Duplicate the |loop| body |factor| - 1 number of times while keeping the loop
 // backedge intact. This will leave the loop with |factor| number of bodies
 // after accounting for the initial body.
 void LoopUnrollerUtilsImpl::Unroll(ir::Loop* loop, size_t factor) {
-  // If we unroll a loop partially it may not be safe to unroll it further.
+  // If we unroll a loop partially it will not be safe to unroll it further.
+  // This is due to the current method of calculating the number of loop
+  // iterations.
```

Comment:
I guess I have to ask what happens in the loop that makes it unsafe, and could the user have simply written the loop that way in the first place?

Reply:
The step count will be incorrect because the current step count code can only handle a single step instruction and in the unrolled loop it will be a series of step operations. The unroller will check this so the user could write code with multiple step instructions but it won't be unrolled. This should be the same for the partially unrolled loop if it was not marked DontUnroll. Basically since we know the loop doesn't meet the conditions we just mark it as DontUnroll, I thought this was a more explicit way of doing it but it's probably just unnecessary redundancy as it would be checked anyway. I can remove it if that would clear up the code a bit more